### PR TITLE
GH-37895: [C++] Feature: support concatenate recordbatches.

### LINKED ISSUE #37895: [C++] support concatenate recordbatches.
### Describe the enhancement requested

User scenario: When we use acero plan, many smaller batches may be generated through agg and hashjoin. In addition, due to the mpp database, there is data distribution. When there are many segments, each segment data is compared at this time. Small, in order to improve performance, we hope to merge multiple fragmented small batches into one large batch for calculation together.

Therefore, similar to the array concatenate operation, recordbatch also requires

### Component(s)

C++

### PR DESCRIPTION
**Pull request by Light-City**
### Rationale for this change

User scenario: When we use acero plan, many smaller batches may be generated through agg and hashjoin. In addition, due to the mpp database, there is data distribution. When there are many segments, each segment data is compared at this time. Small, in order to improve performance, we hope to merge multiple fragmented small batches into one large batch for calculation together.

### What changes are included in this PR?

record_batch.cc
record_batch.h
record_batch_test.cc

### Are these changes tested?

yes, see record_batch_test.cc

### Are there any user-facing changes?

yes

* Closes: #37895

### CODE REVIEW

File: cpp/src/arrow/record_batch.h
Comment (bkietz) [inline cpp/src/arrow/record_batch.h:353]: Nit: either use "record batches" or "RecordBatch"

File: cpp/src/arrow/record_batch.cc
Comment (bkietz) [inline cpp/src/arrow/record_batch.cc:443]: Instead, please make this deep copy the batch to conform with ConcatenateArrays https://github.com/apache/arrow/issues/37878

File: cpp/src/arrow/record_batch.cc
Comment (bkietz) [inline cpp/src/arrow/record_batch.cc:472]:
```suggestion
  return RecordBatch::Make(std::move(schema), length, std::move(columns));
```

File: cpp/src/arrow/record_batch.cc
Comment (bkietz) [inline cpp/src/arrow/record_batch.cc:467]: Instead of aborting, please raise the error. We have a helper macro for this:
```suggestion
      ARROW_ASSIGN_OR_RAISE(auto array, Concatenate(data, pool));
```

File: cpp/src/arrow/record_batch.cc
Comment (bkietz) [inline cpp/src/arrow/record_batch.cc:468]: Instead, please compute the length explicitly before this loop as in the zero columns case

File: cpp/src/arrow/record_batch.cc
Comment (bkietz) [inline cpp/src/arrow/record_batch.cc:454]: I don't think you need a special case for zero columns:

```suggestion
  // special case: null batch, no data, just length
  for (size_t i = 0; i < batches.size(); ++i) {
    length += batches[i]->num_rows();
  }
```

Then the loop `for (int col = 0; col < cols; ++col)` will just never be entered if `cols == 0`

File: cpp/src/arrow/record_batch_test.cc
Comment (bkietz) [inline cpp/src/arrow/record_batch_test.cc:568]: I think you could use `gen.BatchOf` instead for less boilerplate

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/record_batch.h
Comment (Light-City) [reply cpp/src/arrow/record_batch.h:353]: ok

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:472]: ok

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:467]: Indeed, return value handling is missing

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:454]: no, My pseudocode is below, with the else branch
if (cols == 0) {
    // special case: null batch, no data, just length
    for (size_t i = 0; i < batches.size(); ++i) {
      length += batches[i]->num_rows();
    }
  } else {
   for (int col = 0; col < cols; ++col)
}

File: cpp/src/arrow/record_batch_test.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch_test.cc:568]: This function is really good, I will replace it with

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:443]: Can it be changed to this method?
Within the col loop, the length of col=0 is assigned, and the lengths of the subsequent columns are verified against the first column.

```cpp
if (col == 0) {
   length = array->length();
} else if (length != array->length()) {
   return Status::Invalid(
            "column index ", i, " length is ", array->length(),
            ", does not match the length of previous columns: ", length);
}
```

Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:443]: For a single batch, I removed this processing and let it follow the normal logic below

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:454]: The reason why the special case of column 0 is processed here is that if we want to implement count(*), then this RecordBatch does not store data, that is, only the length. This scenario is triggered when small batches need to be merged.

Comment (js8544) [reply cpp/src/arrow/record_batch.cc:451]: Will the following change make any difference on the result?
```cpp
Result<std::shared_ptr<RecordBatch>> ConcatenateRecordBatches(
    const RecordBatchVector& batches, MemoryPool* pool) {
  int64_t length = 0;
  size_t n = batches.size();
  if (n == 0) {
    return Status::Invalid("Must pass at least one recordbatch");
  }
  int cols = batches[0]->num_columns();
  auto schema = batches[0]->schema();
  for (int i = 0; i < batches.size(); ++i) {
    length += batches[i]->num_rows();
    if (!schema->Equals(batches[i]->schema())) {
      return Status::Invalid(
          "Schema of RecordBatch index ", i, " is ", batches[i]->schema()->ToString(),
          ", which does not match index 0 recordbatch schema: ", schema->ToString());
    }
  }

  std::vector<std::shared_ptr<Array>> concatenated_columns;
  for (int col = 0; col < cols; ++col) {
    ArrayVector column_arrays;
    for (const auto & batch : batches) {
      column_arrays.emplace_back(std::move(batch->column(col)));
    }
    ARROW_ASSIGN_OR_RAISE(auto concatenated_column, Concatenate(column_arrays, pool))
    concatenated_columns.emplace_back(std::move(concatenated_column));
  }
  return RecordBatch::Make(std::move(schema), length, std::move(concatenated_columns));
}
```
I assumed:
1. Schema checking can be moved out of the nested loop.
2. The result length is just the sum of all batch lengths, no matter how many (or zero) columns there are.
3. Each column has the same length in a RecordBatch, so no checking of column length is needed here.
4. Variable names like `column_data`, `data`, `columns` are ambiguous so I changed them to more descriptive names.

File: cpp/src/arrow/record_batch_test.cc
Comment (js8544) [inline cpp/src/arrow/record_batch_test.cc:584]: It's desirable to have tests on edge cases such as zero-length RecordBatch and zero-column ones.

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc]: The code looks very refreshing, but it does not handle the case of cols = 0. In this case, cols = 0, length is not obtained.

For a batch, there may be cols = 0, the schema is empty, but there is length. This kind of batch is used to process count(*) and does not store the actual columns, only the number of rows.

Comment (js8544) [reply cpp/src/arrow/record_batch.cc]:
```cpp
for (int i = 0; i < batches.size(); ++i) {
    length += batches[i]->num_rows();
    ...
}
```
Doesn't this include the cols=0 case?

Comment (Light-City) [reply cpp/src/arrow/record_batch.cc]: Sorry, you are correct, it is handled here

Comment (Light-City) [reply cpp/src/arrow/record_batch.cc]: Thanks for the modification. This code modification has been submitted.
 note:
```
column_arrays.emplace_back(std::move(batch->column(col)));
```
Change to
``` 
column_arrays.emplace_back(batch->column(col));
```

File: cpp/src/arrow/record_batch_test.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch_test.cc:584]: yes, i add this test.

**Review by js8544 — APPROVED** [review]
+1. Thanks! But it will need @bkietz's approval before merging.

**Comment by js8544** [comment]
I think the C++ Windows Checks are failing due to this change but I couldn't figure out why...

**Comment by Light-City** [comment]
> I think the C++ Windows Checks are failing due to this change but I couldn't figure out why...

+1, I am also very confused, mac and linux are compiled correctly

But Windows is

```

[345/901] Linking CXX executable release\arrow-table-test.exe
FAILED: release/arrow-table-test.exe 
cmd.exe /C "cd . && D:\a\_temp\msys64\clang64\bin\c++.exe -Qunused-arguments -fcolor-diagnostics  -Wa,-mbig-obj -Wall -Wextra -Wdocumentation -Wshorten-64-to-32 -Wno-missing-braces -Wno-unused-parameter -Wno-constant-logical-operand -Wno-return-stack-address -Wdate-time -Wno-unknown-warning-option -Wno-pass-failed -mxsave -msse4.2  -O3 -DNDEBUG -O2  @CMakeFiles\arrow-table-test.rsp -o release\arrow-table-test.exe -Wl,--out-implib,release\libarrow-table-test.dll.a -Wl,--major-image-version,0,--minor-image-version,0  && cd ."
ld.lld: error: undefined symbol: arrow::ConcatenateRecordBatches(std::__1::vector<std::__1::shared_ptr<arrow::RecordBatch>, std::__1::allocator<std::__1::shared_ptr<arrow::RecordBatch>>> const&, arrow::MemoryPool*)
>>> referenced by src/arrow/CMakeFiles/arrow-table-test.dir/record_batch_test.cc.obj:(arrow::TestRecordBatch_ConcatenateRecordBatches_Test::TestBody())
>>> referenced by src/arrow/CMakeFiles/arrow-table-test.dir/record_batch_test.cc.obj:(arrow::TestRecordBatch_ConcatenateRecordBatches_Test::TestBody())
```

**Comment by Light-City** [comment]
> I think the C++ Windows Checks are failing due to this change but I couldn't figure out why...

Found that ARROW_EXPORT is missing, this should be the reason

File: cpp/src/arrow/record_batch.cc
Comment (mapleFU) [inline cpp/src/arrow/record_batch.cc:454]: should we call `reserve` here?

File: cpp/src/arrow/record_batch.cc
Comment (mapleFU) [inline cpp/src/arrow/record_batch.cc:456]: ditto

File: cpp/src/arrow/record_batch_test.cc
Comment (mapleFU) [inline cpp/src/arrow/record_batch_test.cc:575]: Would you mind test content equals `{b1 concat b2}` ? (using `Slice` or `Concatenate` )

Comment (Light-City) [reply cpp/src/arrow/record_batch_test.cc:575]: ok, I've added it

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:457]: done

File: cpp/src/arrow/record_batch.cc
Comment (Light-City) [reply cpp/src/arrow/record_batch.cc:454]: ok

**Review by mapleFU — APPROVED** [review]

File: cpp/src/arrow/record_batch.h
Comment (mapleFU) [inline cpp/src/arrow/record_batch.h:357]: Nit: can we port some comment from `ConcatenateTables`? (And we don't have a `ConcatenateTablesOptions` and force schema equal)

```c++
/// \brief Construct a new table from multiple input tables.
///
/// The new table is assembled from existing column chunks without copying,
/// if schemas are identical. If schemas do not match exactly and
/// unify_schemas is enabled in options (off by default), an attempt is
/// made to unify them, and then column chunks are converted to their
/// respective unified datatype, which will probably incur a copy.
/// :func:`arrow::PromoteTableToSchema` is used to unify schemas.
///
/// Tables are concatenated in order they are provided in and the order of
/// rows within tables will be preserved.
///
/// \param[in] tables a std::vector of Tables to be concatenated
/// \param[in] options specify how to unify schema of input tables
/// \param[in] memory_pool MemoryPool to be used if null-filled arrays need to
/// be created or if existing column chunks need to endure type conversion
/// \return new Table
```

Comment (Light-City) [reply cpp/src/arrow/record_batch.h:361]: yes, I add some comment.

**Review by bkietz — APPROVED** [review]

**Comment by bkietz** [comment]
CI failure seems unrelated
+1

**Pull request merged**
